client/js/lounge.js

```diff
@@ -52,6 +52,84 @@ $(function() {
 
 	var favicon = $("#favicon");
 
+	var emoji = new window.EmojiConvertor();
+	emoji.replace_mode = "unified";
+	emoji.replace_colons("");  // this is needed to create the colons map
+	var emojies = Object.keys(emoji.map.colons);
+
+	// Autocompletion Strategies
+
+	var emojiStrategy = {
+		id: "emoji",
+		match: /\B:([-+\w]*)$/,
+		search: function(term, callback) {
+			callback($.map(emojies, function(e) {
+				return e.indexOf(term) === 0 ? e : null;
+			}));
+		},
+		template: function(value) {
+			return emoji.replace_colons(":" + value + ":") + " " + value;
+		},
+		replace: function(value) {
+			return emoji.replace_colons(":" + value + ":") + " ";
+		},
+		index: 1
+	};
+
+	var nicksStrategy = {
+		id: "nicks",
+		match: /\B@([a-zA-Z_[\]\\^{}|`@][a-zA-Z0-9_[\]\\^{}|`-]*)$/,
```

Comment:
@yashsriv I don't want the autocomplete popup to appear when just typing names. The old feature needs to work exactly the same as before.

2 use cases:

1. Tab complete. Work exactly as before.

Type characters, press <tab>, get autofilled with rest of username. No popup, no nothing.

2. "@", work exactly as you have it here.

Press "@" and a letter, get popup with list of usernames, gets filtered as you type letters.

The old one you had, where the autocomplete would pop up when you just type a few letters of a username was far too intrusive and would be horrible.

As I said in my overall comment, if you can't figure out a nice way to keep both of these use cases, just remove the username autocomplete from this PR and re-enable tab-complete, and we'll figure it out in another PR. I'd rather another PR than this one takes months to get in. Your work is pretty good and I would love to get this in sooner rather than later. Let's not feature-creep it. We've got a good structure for adding these, which is what @xPaw wanted when he suggested adding users to this.